cmd/occollector/app/builder/receivers_builder.go

```diff
@@ -0,0 +1,273 @@
+// Copyright 2019, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package builder
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-service/consumer"
+	"github.com/open-telemetry/opentelemetry-service/internal"
+	"github.com/open-telemetry/opentelemetry-service/internal/configmodels"
+	"github.com/open-telemetry/opentelemetry-service/internal/factories"
+	"github.com/open-telemetry/opentelemetry-service/processor/multiconsumer"
+	"github.com/open-telemetry/opentelemetry-service/receiver"
+)
+
+// builtReceiver is a receiver that is built based on a config. It can have
+// a trace and/or a metrics component.
+type builtReceiver struct {
+	trace   receiver.TraceReceiver
+	metrics receiver.MetricsReceiver
+}
+
+// Stop the receiver.
+func (rcv *builtReceiver) Stop() error {
+	var errors []error
+	if rcv.trace != nil {
+		err := rcv.trace.StopTraceReception(context.Background())
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	if rcv.metrics != nil {
+		err := rcv.metrics.StopMetricsReception(context.Background())
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	return internal.CombineErrors(errors)
+}
+
+// Start the receiver.
+func (rcv *builtReceiver) Start(asyncErrorChan chan<- error) error {
+	var errors []error
+	if rcv.trace != nil {
+		err := rcv.trace.StartTraceReception(context.Background(), asyncErrorChan)
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	if rcv.metrics != nil {
+		err := rcv.metrics.StartMetricsReception(context.Background(), asyncErrorChan)
+		if err != nil {
+			errors = append(errors, err)
+		}
+	}
+
+	return internal.CombineErrors(errors)
+}
+
+// Receivers is a map of receivers created from receiver configs.
+type Receivers map[configmodels.Receiver]*builtReceiver
+
+// StopAll stops all receivers.
+func (rcvs Receivers) StopAll() {
+	for _, rcv := range rcvs {
```

Comment:
let's do it with goroutines in order to reduce shutdown time.

Reply:
I suggest that we keep it as is. We can move to goroutines if we see that it is actually taking noticeable time. I'd like to have evidence for that before making the code more complex.

Also, we can probably do the same for StartAll(), but again only after we see that it is really slow.

Reply:
Typically we have a grace period to shutdown, so stopping reception asap is important to minimize chances of data loss (we still have to flush the whole pipeline) and redirect clients to other instances. StartAll in that sense is not as important to be concurrent. Be aware that if in the core we just have receivers that close fast we will never see that as an issue. That said it is fine to do in at a later step since we also need a timeout for the operation. Please, open an issue to track that.

Reply:
OK, let's revisit in separate PR.